core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java

```diff
@@ -232,7 +232,14 @@ public boolean isCollectingTaskNeeded(QueryId queryId, PlanFragment plan)
             return false;
         }
 
-        return !getSourceStageInnerLazyDynamicFilters(plan).isEmpty();
+        if (!plan.getPartitioning().equals(SOURCE_DISTRIBUTION)) {
+            return false;
+        }
+
+        // dynamic filters produced within source stage are required by current stage to progress
+        // either directly (e.g. consumed by table scan) or indirectly (e.g. are required by join build
+        // source stage in order to complete).
+        return !getReplicatedDynamicFilters(plan.getRoot()).isEmpty();
```

Comment:
I don't understand this change at all.  Someone else should review it.

Reply:
@raunaqmorarka can you describe this in more detail so that we can validate that this change is okay tomorrow?

Reply:
The change essentially does: If this is a source stage and some dynamic filters is produced by this stage (for broadcast join), then we need extra DF collecting task.

I've updated comment here:
```
            // Dynamic filters can be collected by single task only for source distribution (non-fixed) stage.
            // Such stage must only contain replicated joins.
```